crates/federate/src/worker.rs

```diff
@@ -0,0 +1,267 @@
+        break;
+      }
+      if (Utc::now() - self.last_state_insert) > save_state_every {
+        self.save_and_send_state(pool).await?;
+      }
+      self.update_communities(pool).await?;
+    }
+    // final update of state in db
+    self.save_and_send_state(pool).await?;
+    Ok(())
+  }
+
+  async fn initial_fail_sleep(&mut self) -> Result<()> {
+    // before starting queue, sleep remaining duration if last request failed
+    if self.state.fail_count > 0 {
+      let elapsed = (Utc::now() - self.state.last_retry).to_std()?;
+      let required = retry_sleep_duration(self.state.fail_count);
+      if elapsed >= required {
+        return Ok(());
+      }
+      let remaining = required - elapsed;
+      tokio::select! {
+        () = sleep(remaining) => {},
+        () = self.stop.cancelled() => {}
+      }
+    }
+    Ok(())
+  }
+  async fn loop_batch(&mut self, pool: &mut DbPool<'_>) -> Result<()> {
+    let latest_id = get_latest_activity_id(pool).await?;
+    let mut id = self.state.last_successful_id;
+    if id == latest_id {
+      // no more work to be done, wait before rechecking
+      tokio::select! {
+        () = sleep(Duration::from_secs(10)) => {},
+        () = self.stop.cancelled() => {}
+      }
+      return Ok(());
+    }
+    let mut processed_activities = 0;
+    while id < latest_id
+      && processed_activities < CHECK_SAVE_STATE_EVERY_IT
+      && !self.stop.is_cancelled()
+    {
+      id += 1;
+      processed_activities += 1;
+      let Some(ele) = get_activity_cached(pool, id).await? else {
+        self.state.last_successful_id = id;
+        continue;
+      };
+      self.send_retry_loop(pool, &ele.0, &ele.1).await?;
+      if self.stop.is_cancelled() {
+        return Ok(());
+      }
+      // send success!
+      self.state.last_successful_id = id;
+      self.state.fail_count = 0;
+    }
+    Ok(())
+  }
+
+  /** this function will only return if (a) send succeeded or (b) worker cancelled */
+  async fn send_retry_loop(
+    &mut self,
+    pool: &mut DbPool<'_>,
+    activity: &SentActivity,
+    object: &SharedInboxActivities,
+  ) -> Result<()> {
+    let inbox_urls = self.get_inbox_urls(activity);
+    if inbox_urls.is_empty() {
+      self.state.last_successful_id = activity.id;
+      return Ok(());
+    }
+    let Some(actor_apub_id) = &activity.actor_apub_id else {
+      return Ok(()); // activity was inserted before persistent queue was activated
+    };
+    let actor = get_actor_cached(pool, activity.actor_type, actor_apub_id).await?;
+
+    let inbox_urls = inbox_urls.into_iter().collect();
+    let requests = prepare_raw(object, actor.as_ref(), inbox_urls, &self.context)
+      .await
+      .into_anyhow()?;
+    for task in requests {
+      // usually only one due to shared inbox
+      let mut req = sign_raw(&task, &self.context, REQWEST_TIMEOUT).await?;
+      tracing::info!("sending out {}", task);
+      while let Err(e) = send_raw(&task, &self.context, req).await {
```

Comment:
It's possible, but mostly no. The code in ap-fed specifically handles most errors as non-retryable (returning success). See:

https://github.com/LemmyNet/activitypub-federation-rust/blob/b0547e7793f2e25c27fa9d292cde342ecb7812e1/src/activity_queue.rs#L186-L218

Basically, my code relies on there being clean seperation between retryable vs permanent errors. If the receiving end returns with 2XX or rejects with 4XX, the sending code moves on. If it rejects with timeout or 5XX, it will be stuck in a loop. It's what I mentioned as disadvantage (1), this code assumes that every activity has a fully correlated chance of returning 5XX or timeout. If one activity causes the receiving end to 5XX or timeout and others *would* succeed, then there's an issue. I don't expect this to happen, but if it does it will be easy to debug.